include/xtensor/xmath.hpp

```diff
@@ -2081,26 +2081,27 @@ namespace detail {
         }
     }
 
-    template <class E, class D, class EVS = DEFAULT_STRATEGY_REDUCERS,
+    template <class T = void, class E, class D, class EVS = DEFAULT_STRATEGY_REDUCERS,
               XTL_REQUIRES(is_reducer_options<EVS>, std::is_integral<D>)>
     inline auto variance(E&& e, D const& ddof, EVS es = EVS())
     {
         decltype(auto) sc = detail::shared_forward<E>(e);
-        return detail::mean_noaxis<void>(square(abs(sc - mean(sc, es))), ddof, es);
+        return detail::mean_noaxis<T>(square(sc - mean<T>(sc, es)), ddof, es);
     }
 
-    template <class E, class D=int, class EVS = DEFAULT_STRATEGY_REDUCERS,
+    template <class T = void, class E, class D=int, class EVS = DEFAULT_STRATEGY_REDUCERS,
               XTL_REQUIRES(is_reducer_options<EVS>)>
     inline auto variance(E&& e, EVS es = EVS())
     {
-        return variance(std::forward<E>(e), 0u, es);
+        return variance<T>(std::forward<E>(e), 0u, es);
     }
 
-    template <class E, class EVS = DEFAULT_STRATEGY_REDUCERS,
+    template <class T = void, class E, class EVS = DEFAULT_STRATEGY_REDUCERS,
               XTL_REQUIRES(is_reducer_options<EVS>)>
     inline auto stddev(E&& e, EVS es = EVS())
     {
-        return sqrt(variance(std::forward<E>(e), es));
+        using result_type = std::conditional_t<std::is_same<T, void>::value, double, T>;
+        return cast<result_type>(sqrt(variance<T>(std::forward<E>(e), es)));
```

Comment:
We need to provide a specialization of cast that implements a SIMD access operator when the `return_type` and the `value_type` of of the expression are the same. Otherwise we might have a serious performance hit here.

This can be done in a dedicated PR.

Reply:
Thanks for the explaining! I also noticed that `stddev` is super slow. Do you have an example implementation somewhere else in the repo?

Reply:
Not really, simd operators of functors are either always implemented, or not at all, but we don't have partial implementation of these operators.